### ISSUE TITLE
Can mask rcnn be trained in batch size == 1?

### ISSUE THREAD

**Issue by mks0601** [facebookresearch/Detectron#211]
I'm implementing mask rcnn on my own for the keypoint detection on the pytorch. As far as I know, different image sizes in mini-batch is not allowed in pytorch, so I`m training the model with batch size == 1. Have you ever successfully trained the mask rcnn in mini-batch size == 1? 

Also, how did you handle the different image sizes in mini-batch? Does caffe2 support this feature?

**Comment by ir413** [comment]
Hi @mks0601 - yes, you can train Mask R-CNN with 1 image per GPU (mini-batch of 1 in case of 1 GPU). For example, the ResNeXt models from our [model zoo](https://github.com/facebookresearch/Detectron/blob/master/MODEL_ZOO.md#fast--mask-r-cnn-baselines-using-precomputed-rpn-proposals) are trained with 1 image per GPU. Note that this will impact the mini-batch size and will require adjusting the learning schedule, which can be done using the "linear scaling rule" (see [this page](https://github.com/facebookresearch/Detectron/blob/2f8161edc3092b0382cab535c977a180a8b3cc4d/GETTING_STARTED.md#2-multi-gpu-training) for more info).

See [this function](https://github.com/facebookresearch/Detectron/blob/master/lib/utils/blob.py#L40) for how lists of images are converted to batch inputs. You could do the same in pytroch. Related to this, you may also want to check out the [aspect-ratio grouping](https://github.com/facebookresearch/Detectron/blob/master/lib/roi_data/loader.py#L139) strategy for saving memory.

**Comment by mks0601** [comment]
Thank you @ir413 

I got to know how you guys handle the mini-batch generation issue. Basically, zero padding with aspect-ratio grouping strategy are used for mini-batch creation.

However, I went through the code and still have a question about the mini-batch processing.  Let me correct if I got into a wrong way. It seems the RPN takes a single image and generate proposals, followed by fast rcnn head and keypoint head. The 2 images mini-batch are fed to the pre-trained ResNet and it outputs feature maps from each conv block in case of FPN implementation. Then, the feature maps of 'each image' is fed to the RPN followed by fast rcnn and keypoint rcnn head. Is it correct?

As the BN layer is only positioned in ResNet part, mini-batch forwarding is only required in that part. The later parts can mimic mini-batch effect by forwarding each single image from the mini-batch and accumulating gradient, followed by normalizing it. Is it correct?

**Comment by ir413** [comment]
Hi @mks0601, glad that's resolved. Closing this since the original issue has been addressed.

The concept of mini-batch in the Fast/Faster/Mask R-CNN models differs a bit from the classification setting since different mini-batch sizes are used for different parts of the model. Namely, the network backbone uses a mini-batch size of [`TRAIN.IMS_PER_BATCH`](https://github.com/facebookresearch/Detectron/blob/master/lib/core/config.py#L90-L92) images (per GPU) while the network heads use [`TRAIN.BATCH_SIZE_PER_IM`](https://github.com/facebookresearch/Detectron/blob/master/lib/core/config.py#L94-L98) RoIs (per image).  

BN layers are replaces with affine transformation layers when fine-tuning ImaneNet pretrained classification models for detection on COCO.

Please refer to the original papers for description of the methods and open a separate issue if you require any further clarification.
